src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -452,6 +461,37 @@ private long writeRawLogTo(AnnotatedLargeText<?> text, long start, OutputStream
         }
     }
 
+    @GuardedBy("completed")
+    private transient LoadingCache<FlowNode,List<String>> nestingCache;
+    private @Nonnull List<String> getNesting(FlowNode node) {
+        // TODO could also use FlowScanningUtils.fetchEnclosingBlocks(node) but this would not let us cache intermediate results
+        synchronized (completed) {
+            if (nestingCache == null) {
+                nestingCache = CacheBuilder.newBuilder().weakKeys().build(new CacheLoader<FlowNode,List<String>>() {
+                    @Override public @Nonnull List<String> load(FlowNode node) {
+                        if (node instanceof BlockEndNode) {
+                            return getNesting(((BlockEndNode) node).getStartNode());
+                        } else {
+                            List<FlowNode> parents = node.getParents();
+                            if (parents.isEmpty()) { // FlowStartNode
+                                return Collections.emptyList();
+                            }
+                            List<String> parent = getNesting(parents.get(0)); // multiple parents is only for BlockEndNode after parallel
+                            if (node instanceof BlockStartNode) {
```

Comment:
@jglick Consider using https://github.com/jenkinsci/workflow-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/SimpleChunkVisitor.java  to collect ForkScanner results - you can easily roll up the result with a simple ArrayDequeue of blocks.